planner/core/optimizer.go

```diff
@@ -120,6 +121,68 @@ func CheckPrivilege(activeRoles []*auth.RoleIdentity, pm privilege.Manager, vs [
 	return nil
 }
 
+// VisitInfo4PrivCheck generates privilege check infos because privilege check of local temporary tables is different
+// with normal tables. `CREATE` statement needs `CREATE TEMPORARY TABLE` privilege from the database, and subsequent
+// statements do not need any privileges.
+func VisitInfo4PrivCheck(is infoschema.InfoSchema, node ast.Node, vs []visitInfo) (privVisitInfo []visitInfo) {
+	if node == nil {
+		return vs
+	}
+
+	switch stmt := node.(type) {
+	case *ast.CreateTableStmt:
+		privVisitInfo = make([]visitInfo, 0, len(vs))
+		for _, v := range vs {
+			if stmt.TemporaryKeyword == ast.TemporaryLocal && v.privilege == mysql.CreatePriv {
+				// `CREATE TEMPORARY TABLE` privilege is required from the database, not the table.
+				newVisitInfo := v
+				newVisitInfo.privilege = mysql.CreateTMPTablePriv
+				newVisitInfo.table = ""
+				privVisitInfo = append(privVisitInfo, newVisitInfo)
+			} else {
+				privVisitInfo = append(privVisitInfo, v)
```

Comment:
Do we need `needCheckTmpTablePriv` here?

Reply:
I don't think so. Any examples?

Reply:
For example: `create table t1 from select * from tmp` . Tough we do not support this syntax currently, adding some checks here can avoid hidden problems in the future.

Reply:
Addressed.